File: Hard/2025-12-31-1970-Last-Day-Where-You-Can-Still-Cross/solution.py
class Solution:
    def canCross(self, row, col, cells, day):
        grid = [[0] * col for _ in range(row)]
        for i in range(day):
            r, c = cells[i]
            grid[r - 1][c - 1] = 1
        
        visited = [[False] * col for _ in range(row)]
        queue = collections.deque()
        
        for c in range(col):
            if grid[0][c] == 0:
                queue.append((0, c))
                visited[0][c] = True
        
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        
        while queue:
            r, c = queue.popleft()
            if r == row - 1:
                return True
            
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < row and 0 <= nc < col and not visited[nr][nc] and grid[nr][nc] == 0:
                    visited[nr][nc] = True
                    queue.append((nr, nc))
        
        return False

    def latestDayToCross(self, row: int, col: int, cells: List[List[int]]) -> int:
        left, right = 1, len(cells)
        answer = 0
        
        while left <= right:
            mid = (left + right) // 2
            if self.canCross(row, col, cells, mid):
                answer = mid
                left = mid + 1
            else:
                right = mid - 1
        
        return answer

File: Hard/2025-12-31-1970-Last-Day-Where-You-Can-Still-Cross/solution.java
class Solution {
    public int latestDayToCross(int row, int col, int[][] cells) {
        int left = 0, right = cells.length - 1;
        while (left < right) {
            int mid = left + (right - left + 1) / 2;
            if (canCross(row, col, cells, mid)) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }

    private boolean canCross(int row, int col, int[][] cells, int day) {
        boolean[][] grid = new boolean[row][col];
        for (int i = 0; i < day; i++) {
            int r = cells[i][0] - 1;
            int c = cells[i][1] - 1;
            grid[r][c] = true;
        }
        
        boolean[] visited = new boolean[col];
        for (int j = 0; j < col; j++) {
            if (!grid[0][j]) {
                visited[j] = true;
            }
        }

        for (int i = 0; i < row; i++) {
            boolean[] newVisited = new boolean[col];
            for (int j = 0; j < col; j++) {
                if (visited[j] && !grid[i][j]) {
                    newVisited[j] = true;
                    if (i == row - 1) {
                        return true;
                    }
                    if (j > 0 && !grid[i][j - 1]) {
                        newVisited[j - 1] = true;
                    }
                    if (j < col - 1 && !grid[i][j + 1]) {
                        newVisited[j + 1] = true;
                    }
                    if (i < row - 1 && !grid[i + 1][j]) {
                        newVisited[j] = true;
                    }
                }
            }
            visited = newVisited;
        }
        return false;
    }
}